### DISCUSSION TITLE
Uncaught TypeError: "Converting circular structure" with @entangled + queryString

### DISCUSSION THREAD

**Discussion by wrabit** [livewire/livewire#3075] [7. Bugs (archived)]
### Description
Full page component, using an entangled value to keep state of active tab. Browsing back and forth triggers:

```
Uncaught TypeError: Converting circular structure to JSON
    --> starting at object with constructor 'Object'
    --- property 'defer' closes the circle
    at JSON.stringify (<anonymous>)
    at Connection.value (index.js:28)
    at sendMessage (index.js:212)
    at Component.value (index.js:222)
    at later (debounce.js:8)
```

### Exact steps to reproduce

1. Create a set of tabs with alpine and livewire
2. Use entangle to record the active tab state
3. Set this property as $queryString
4. Make a few clicks forward, then go back with browser history button a few times.
5. Check console

### Stripped-down, copy-pastable code snippets

Available as repo: Repo: https://github.com/wrabit/lw-bugs 

or, this is all that has been added to vanilla setup:

#### routes.php
```
Route::get('tabs', Tabs::class);
```

#### Http\Livewire\Tabs
```
namespace App\Http\Livewire;

use Livewire\Component;

class Tabs extends Component
{
    public $step = 1;

    protected $queryString = ['step'];

    public function render()
    {
        return view('livewire.tabs', ['step' => $this->step])->layout('layouts.base');
    }
}
```

#### resources/views/livewire/tabs.blade.php
```
<div x-data="{
        step: @entangle('step'),
    }">
    <ul>
        <li><a href="#" @click.prevent="step = 1">1</a></li>
        <li><a href="#" @click.prevent="step = 2">2</a></li>
        <li><a href="#" @click.prevent="step = 3">3</a></li>
    </ul>
    
    <div>
        <div x-show="step == 1">Step 1</div>
        <div x-show="step == 2">Step 2</div>
        <div x-show="step == 3">Step 3</div>
    </div>
</div>
```

#### layouts/app.php
```
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">
<head>
    @livewireStyles
    <script src="https://cdn.jsdelivr.net/gh/alpinejs/alpine@v2.x.x/dist/alpine.min.js" defer></script>
</head>
<body>
{{ $slot  }}
@livewireScripts
</body>
</html>
```

**Context**
- Livewire version: checked on 2.3 -> master
- Laravel version: checked on 8.0 -> 8.4
- Alpine version: Latest 2.x.x
- Browser: Chrome

**Comment by joshhanley** [comment]
@calebporzio I can confirm, I have tested this, haven't looked into what is causing it yet. 

A more reliable way to trigger the error is to click on tab 2, refresh the page, then click the back button in the browser.
The query string changes back to `step=1` but the circular reference error occurs.

I have pushed up my test repo have a look at this branch "circularIssue" here, just composer install and then go to base route. 
https://github.com/joshhanley/LivewirePlayground/tree/circularIssue

**Comment by lao9s** [comment]
Same problem.

**Comment by ryatkins** [comment]
I had this issue when trying to reorder items on a page that use entangle on an input field.

I fixed it by adding wire:key to my input element.

**Comment by flatcapco** [comment]
I get this error with an entangled datepicker that is conditionally shown from another livewire select compontent.

The error doesn't happen if I select the option that reveals this field first. 
However if I select 2 other conditional options first from the select then this field it throws the error.

** wire:key did not fix it in my case as its not within a loop

**Comment by ryatkins** [comment]
Without seeing your code, two things you could try:
1. Wrap the item you are selectively removing in an outer DIV so that DIV is also present on the page.

```
<div>
@if($condition)
   <x-datepicker />
@endif
</div>
```

or

2. Hide the datepicker with x-show using Alpine

**Comment by wrabit** [comment]
@latwelve + @ryatkins these are both different issues which I am aware of, but this issue is about browsing back and forth

**Comment by n3b0r** [comment]
Same here, if it's needed I can provide a full example.

**Comment by phuclh** [comment]
Any solution for this?

**Comment by IP-Developer** [comment]
+1 on this.
In my case, I set the entanglement on x-init for an object (associative array in php), and then I have to manually call `this.$wire.set('customer', this.customer)` which results in this error.

**Comment by p-andrey** [comment]
Same problem when use @entangle()

**Comment by ph7jack** [comment]
This problem still persists

**Comment by glspdotnet** [comment]
I just ran into the same issue and figured that this might help someone else.

In my case the issue seemed to be the need for static HTML content _next to_ the element that was `@entangle()`d.

As @ryatkins mentioned, a wrapper `<div>` solves the issue, but I was also able to eliminate the problem without modifying my document structure by simply inserting an HTML comment between the conditionally-displayed element and the entangled element, like so:
```blade
<div>
    Persistent content here.
</div>
@isset ($livewirePropertyOne)
    <div>
        Some conditional content with no other Livewire/Alpine functionality here.
    </div>
@endisset
<!-- Static DOM Node to assist Livewire's diffing algorithm -->
<div
    x-data="{ alpineProp: @entangle('livewirePropertyTwo') }"
    x-on:some-event="alpineProp = $event.target.value"
>
    More content here.
</div>
```
@wrabit I cloned your repo and was only able to reproduce the issue once, after hitting my browser's back button enough times to bring me back to my new tab page. I wasn't able to reproduce it again, unfortunately.

**Comment by Pulstarter** [comment]
I am running into this issue with Alpine 2.8.2 when I have several selectboxes in cascade, and one of them filter the others. If I apply @glspdotnet  solution, it only works for the first time, but when I filter a second time, then it throws the error.
